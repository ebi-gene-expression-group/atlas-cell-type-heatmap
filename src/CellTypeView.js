import React from "react"
import PropTypes from 'prop-types'

import URI from 'urijs'
import _ from "lodash"

import CellTypeHighchartsHeatmap from "./CellTypeHighchartsHeatmap"
import CalloutAlert from './CalloutAlert'
import LoadingOverlay from './LoadingOverlay'
import DataSeriesHeatmapLegent from './DataSeriesHeatmapLegend'


class CellTypeView extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      expressionByCellTypeData: [],
      solrResponse: [],
      filteredData: [],
      experimentAccessionsBySpecies: [],
      responseBySpecies: [],
      axisData: {},
      isLoading: false,
      speciesList: [],
      selectedSpecies: null
    }
  }

  async _fetchExpressionDataForHeatmap({resource, host}) {
    this.setState({
      isLoading: true
    })

    const url = URI(resource, host).toString()

    try {
      const response = await fetch(url)

      if (!response.ok) {
        throw new Error(`${url} => ${response.status}`)
      }

      const { species } = this.props

      let results = Object.values(await response.json())[0][species]

      const experimentAccessions = Object.keys(results)
      const markerGenes = Object.keys(results[experimentAccessions[0]].markerGeneExpression).sort()
      const technologyType = Object.values(results).map(experimentAccession => experimentAccession.technologyType)

      this.setState({
        expressionByCellTypeData: response,
        responseBySpecies: results,
        axisData: this._heatmapAxis(markerGenes, results, technologyType),
        heatmapData: this._factorHeatmapData(markerGenes, results),
        filteredData: this._factorHeatmapData(markerGenes, results),
        isLoading: false,
        selectedSpecies: null
      })

    } catch(e) {
      this.setState({
        data: null,
        isLoading: false,
        hasError: {
          description: `There was a problem communicating with the server. Please try again later.`,
          name: e.name,
          message: e.message
        }
      })
    }
  }

  _factorHeatmapData(markerGenesBySpecies, responseBySpecies) {

    const experimentAccessions = Object.keys(responseBySpecies)

    // const cellTypes = experimentAccessions.map(accession =>
    //   Object.keys(responseBySpecies[accession][Object.keys(responseBySpecies[accession])[0]])
    // )

    const cellTypes = Object.keys(Object.values((Object.values(responseBySpecies)[0]).markerGeneExpression)[0])

    const heatmapData = markerGenesBySpecies.map(markerGene =>
      _.flatMap(
        experimentAccessions.map((experimentAccession) =>
          cellTypes.map(cellType => {
            return (responseBySpecies[experimentAccession]).markerGeneExpression[markerGene][cellType]
          })
        )
      )
    )
    return heatmapData
  }

  _heatmapAxis(markerGenesBySpecies, responseBySpecies, technologyType) {
    const experimentAccessions = Object.keys(responseBySpecies)
    const cellType = Object.keys(Object.values((Object.values(responseBySpecies)[0]).markerGeneExpression)[0])[0]

    const yAxisCategory = _.flatMap(
      experimentAccessions.map((experimentAccession, idx) =>
        ({
          cellType: cellType.substring(1, cellType.length - 1),
          experimentAccession: experimentAccession,
          technologyType: technologyType[idx][0]
        })
      )
    )
    return ({ y: markerGenesBySpecies, x: yAxisCategory })
  }

  componentDidMount() {
    this._fetchExpressionDataForHeatmap(this.props)
  }

  componentDidUpdate(previousProps) {
    if (previousProps.resource !== this.props.resource) {
      this._fetchExpressionDataForHeatmap(this.props)
    }
  }

  componentDidCatch(error, info) {
    this.setState({
      hasError: {
        description: `There was a problem rendering this component.`,
        name: error.name,
        message: `${error.message} â€“ ${info}`
      }
    })
  }

  render() {
    const { axisData, hasError, isLoading, filteredData } = this.state
    const { wrapperClassName, plotWrapperClassName, species } = this.props
    const { hasDynamicHeight, defaultHeatmapHeight, heatmapRowHeight } = this.props

    return (
      hasError ?
        <CalloutAlert error={hasError}/> :

        <div className={wrapperClassName}>
          <div className={plotWrapperClassName} style={{position: `relative`}}>
            <CellTypeHighchartsHeatmap
              species={{species}}
              axisData={axisData}
              heatmapData={filteredData}
              chartHeight={defaultHeatmapHeight}
              hasDynamicHeight={axisData.y && axisData.y.length > 5 ? hasDynamicHeight : false} // don't want dynamic height if there is little or no data
              heatmapRowHeight={heatmapRowHeight}
            />
            <LoadingOverlay
              show={isLoading}
            />
          </div>
          <DataSeriesHeatmapLegent />
        </div>
    )
  }
}

CellTypeView.propTypes = {
  host: PropTypes.string.isRequired,
  resource: PropTypes.string.isRequired,
  wrapperClassName: PropTypes.string,
  plotWrapperClassName: PropTypes.string,
  defaultHeatmapHeight: PropTypes.number,
  hasDynamicHeight: PropTypes.bool,
  heatmapRowHeight: PropTypes.number,
  species: PropTypes.string.isRequired
}

CellTypeView.defaultProps = {
  wrapperClassName: `small-8 columns row`,
  plotWrapperClassName: `small-12 columns`,
  defaultHeatmapHeight: 300,
  hasDynamicHeight: true,
  heatmapRowHeight: 20
}

export default CellTypeView
