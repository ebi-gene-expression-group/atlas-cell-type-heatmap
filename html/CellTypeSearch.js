import React from 'react'
import PropTypes from 'prop-types'

import URI from 'urijs'
import CellTypeView from '../src/index.js'
import PlotSettingsDropdown from "../src/PlotSettingsDropdown"
import _ from "lodash"
import LoadingOverlay from "../src/LoadingOverlay"

const names = [`sex`, `organism_part`], values = [[`female`, `male`], [`lymph node`, `pancreas`, `skin`]]

class CellTypeSearch extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      selectedName: `organism_part`,
      selectedValue: `pancreas`,
      specieses: [],
      cellTypes: {}
    }
  }

  async _fetchCellTypes(name, value) {
    this.setState({
      isLoading: true
    })

    const url = `http://localhost:8080/gxa/sc/json/metadata-search/cell-type/name/${name}/value/${value}`

    try {
      const response = await fetch(url)

      if (!response.ok) {
        throw new Error(`${url} => ${response.status}`)
      }

      const result = Object.values(await response.json())[0]
      const specieses = Object.keys(result)
      const cellTypes = specieses.map(species => _.uniq(_.flatMap(Object.values(result[species]))))

      this.setState({
        result: result,
        specieses: specieses,
        cellTypes: cellTypes,
        isLoading: false
      })
    } catch(e) {
      this.setState({
        isLoading: false,
        hasError: {
          description: `There was a problem communicating with the server. Please try again later.`,
          name: e.name,
          message: e.message,
          isLoading: false
        }
      })
    }
  }

  render() {
    const requestParams = URI(location.search).query(true)
    const {selectedName, selectedValue, selectedSpecies, selectedCellType, specieses, cellTypes, isLoading} = this.state

    const valueOptions = values[names.indexOf(selectedName)]
      .map((v) => ({
        value: v.toString(),
        label: v,
        isDisabled: false
      }))

    const nameOptions = names
      .map((v) => ({
        value: v.toString(),
        label: v,
        isDisabled: false
      }))

    const cellTypesOptions = selectedSpecies ?
      cellTypes[specieses.indexOf(selectedSpecies)]
        .map((v) => ({
          value: v.toString().substring(1, v.length - 1),
          label: v.toString().substring(1, v.length - 1),
          isDisabled: false
        }))
        .filter(v => v.value !== `not available` && v.value !== `not applicable` && v.value !== ``) :
      []

    const speciesOptions = specieses
      .map((v) => ({
        value: v.toString(),
        label: v,
        isDisabled: false
      }))

    return (
      <div>
        <div className={`row expanded`}>
          <div className={`small-12 medium-5 columns`}>
            <PlotSettingsDropdown
              labelText={`Characteristic name:`}
              options={nameOptions}
              onSelect={(selectedOption) => {
                this.setState({
                  selectedName: selectedOption.value
                })
              }}
              value={{value: selectedName, label: selectedName}}
            />
          </div>
          <div className={`small-12 medium-5 columns`}>
            <PlotSettingsDropdown
              labelText={`Characteristic value:`}
              options={valueOptions}
              onSelect={(selectedOption) => {
                this.setState({
                  selectedValue: selectedOption.value
                })
              }}
              value={{value: selectedValue, label: selectedValue}}
            />
          </div>
          <div className={`small-12 medium-2 columns`} style={{paddingTop: `25px`}}>
            <button className={`button`} onClick={(event) => {
              event.preventDefault()
              this.props.history.push(`/`)
              this._fetchCellTypes(this.state.selectedName, this.state.selectedValue)
              this.setState({
                selectedSpecies: ``,
                selectedCellType: ``
              })
            }}>
              Search</button>
          </div>
        </div>



        <div className={`row expanded`}>
          <div className={`small-12 medium-5 columns`}>
            <PlotSettingsDropdown
              labelText={`Species:`}
              options={speciesOptions}
              onSelect={(selectedOption) => {
                this.setState({
                  selectedSpecies: selectedOption.value
                })
              }}
              value={{value: selectedSpecies || requestParams.species, label: selectedSpecies || requestParams.species}}
            />
          </div>
          <div className={`small-12 medium-5 columns`}>
            <PlotSettingsDropdown
              labelText={`Inferred cell types:`}
              options={cellTypesOptions}
              onSelect={(selectedOption) => {
                this.setState({
                  selectedCellType: selectedOption.value
                })
              }}
              value={{value: selectedCellType || requestParams.cellType, label: selectedCellType || requestParams.cellType}}
            />
          </div>
          <div className={`small-12 medium-2 columns`} style={{paddingTop: `25px`}}>
            <button className={`button`} onClick={(event) => {
              event.preventDefault()
              this.props.history.push(`/search?species=` + this.state.selectedSpecies + `&cellType=` + this.state.selectedCellType)
              this.setState({
                cellType: this.state.selectedCellType
              })}}>
              Search
            </button>
          </div>


        </div>
        {
          requestParams.cellType && <div style={{paddingBottom: `25px`}}>
            <CellTypeView
              wrapperClassName={`row expanded`}
              resource={`json/metadata-search/expression/name/inferred_cell_type/value/${requestParams.cellType}`}
              host={`http://localhost:8080/gxa/sc/`}
              hasDynamicHeight={true}
              species={requestParams.species}
              heatmapRowHeight={30}
            />
          </div>
        }
        <LoadingOverlay
          show={isLoading}
        />
      </div>
    )
  }
}

CellTypeSearch.propTypes = {
  atlasUrl: PropTypes.string.isRequired,
  history: PropTypes.object.isRequired
  // If we really need to know historyâ€™s propTypes (e.g. for tests) here they are:
  // history: PropTypes.shape({
  //   length: PropTypes.number.isRequired,
  //   action: PropTypes.oneOf([`POP`, `PUSH`, `REPLACE`]).isRequired,
  //   location: PropTypes.shape({
  //     hash: PropTypes.string.isRequired,
  //     key: PropTypes.string.isRequired,
  //     query: PropTypes.string,
  //     state: PropTypes.string
  //   }).isRequired,
  //   createHref: PropTypes.func.isRequired,
  //   push: PropTypes.func.isRequired,
  //   replace: PropTypes.func.isRequired,
  //   go: PropTypes.func.isRequired,
  //   goBack: PropTypes.func.isRequired,
  //   goForward: PropTypes.func.isRequired,
  //   block: PropTypes.func.isRequired,
  //   listen: PropTypes.func.isRequired
  // })
}

export default CellTypeSearch