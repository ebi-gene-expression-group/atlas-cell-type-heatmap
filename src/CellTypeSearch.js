import React from 'react'
import PropTypes from 'prop-types'

import URI from 'urijs'
import CellTypeView from './CellTypeView'
import LoadingOverlay from "./LoadingOverlay"
import PlotSettingsDropdown from "./PlotSettingsDropdown"
import _ from "lodash"

class CellTypeSearch extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      selectedSpecies: ``,
      selectedCellType: ``,
      cellType: ``
    }
  }

  render() {
    const requestParams = URI(location.search).query(true)
    const {speciesList, cellTypePayload, host, resource} = this.props
    const {selectedSpecies, selectedCellType, isLoading} = this.state

    const cellTypeList = speciesList.map(species => _.uniq(_.flatMap(Object.values(cellTypePayload[species]))))

    const cellTypesOptions = selectedSpecies ?
      cellTypeList[speciesList.indexOf(selectedSpecies)]
        .map((v) => ({
          value: v.toString().substring(1, v.length - 1),
          label: v.toString().substring(1, v.length - 1),
          isDisabled: false
        }))
        .filter(v => v.value !== `not available` && v.value !== `not applicable` && v.value !== ``) :
      []

    const speciesOptions = speciesList
      .map((v) => ({
        value: v.toString(),
        label: v,
        isDisabled: false
      }))

    return (
      <div>
        <div className={`row expanded`}>
          <div className={`small-12 medium-5 columns`}>
            <PlotSettingsDropdown
              labelText={`Species:`}
              options={speciesOptions}
              onSelect={(selectedOption) => {
                this.setState({
                  selectedSpecies: selectedOption.value
                })
              }}
              value={{value: selectedSpecies || requestParams.species, label: selectedSpecies || requestParams.species}}
            />
          </div>
          <div className={`small-12 medium-5 columns`}>
            <PlotSettingsDropdown
              labelText={`Inferred cell types:`}
              options={cellTypesOptions}
              onSelect={(selectedOption) => {
                this.setState({
                  selectedCellType: selectedOption.value
                })
              }}
              value={{value: selectedCellType || requestParams.cellType, label: selectedCellType || requestParams.cellType}}
            />
          </div>
          <div className={`small-12 medium-2 columns`} style={{paddingTop: `25px`}}>
            <button className={`button`} onClick={(event) => {
              event.preventDefault()
              this.props.history.push(`/search?species=` + this.state.selectedSpecies + `&cellType=` + this.state.selectedCellType)
              this.setState({
                cellType: this.state.selectedCellType
              })}}>
              Search
            </button>
          </div>

        </div>
        {
          requestParams.cellType && <div style={{paddingBottom: `25px`}}>
            <CellTypeView
              wrapperClassName={`row expanded`}
              resource={`${resource}/${requestParams.cellType}`}
              host={host}
              hasDynamicHeight={true}
              species={requestParams.species}
              heatmapRowHeight={30}
            />
          </div>
        }
        <LoadingOverlay
          show={isLoading}
        />
      </div>
    )
  }
}

CellTypeSearch.propTypes = {
  history: PropTypes.object.isRequired,
  host: PropTypes.string.isRequired,
  resource: PropTypes.string.isRequired,
  speciesList: PropTypes.arrayOf(PropTypes.string).isRequired,
  cellTypeList: PropTypes.arrayOf(PropTypes.string).isRequired
}

export default CellTypeSearch