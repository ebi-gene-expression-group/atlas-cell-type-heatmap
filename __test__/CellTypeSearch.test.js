import React from 'react'
import Enzyme from 'enzyme'
import {shallow, mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import {Route, MemoryRouter} from 'react-router-dom'

import '@babel/polyfill'

import CellTypeSearch from '../src/CellTypeSearch'
import CellTypeView from '../src/CellTypeView'
import LoadingOverlay from '../src/LoadingOverlay'
import PlotSettingsDropdown from '../src/PlotSettingsDropdown'
import renderer from "react-test-renderer"

Enzyme.configure({ adapter: new Adapter() })

describe(`CellTypeSearch`, () => {
  const props = {
    history: {},
    resource: `json/metadata-search/expression/name/inferred_cell_type/value/`,
    host: `http://localhost:8080/gxa/sc/`,
    speciesList: [`Homo sapiens`],
    cellTypePayload: {"Homo sapiens":{"E-MTAB-5061":[`[beta cell]`,`[not available]`,`[pancreatic A cell]`]}}
  }

  beforeAll(() => {
    delete window.location
    window.location = { search: `?species=Homo%20sapiens&cellType=beta%20cell` }
  })

  test(`renders two dropdown boxes and one button for searching`, () => {
    const wrapper = shallow(<CellTypeSearch {...props} />)
    expect(wrapper.find(PlotSettingsDropdown).length).toBe(2)
    expect(wrapper.find(`button`).length).toBe(1)
  })

  test(`renders a loading overlay while loading`, () => {
    const wrapper = shallow(<CellTypeSearch {...props} />)
    expect(wrapper.find(LoadingOverlay).length).toBe(1)
  })

  test(`renders a cell type heatmap after loading`, () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[ `/search?species=Homo%20sapiens&cellType=beta%20cell` ]}>
        <Route
          exact path={`/search`}
          render={
            ({history}) =>
              <CellTypeSearch
                {...props}
                history={history}/>
          } />
      </MemoryRouter>
    )
    expect(wrapper).toContainMatchingElement(CellTypeView)
  })

  test(`with no explicit props matches snapshot`, () => {
    const tree = renderer
      .create(<CellTypeSearch {...props}/>)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

})
