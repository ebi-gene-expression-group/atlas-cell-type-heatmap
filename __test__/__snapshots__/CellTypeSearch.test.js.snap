// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellTypeSearch with no explicit props matches snapshot 1`] = `
<div>
  <div
    className="row expanded"
  >
    <div
      className="small-12 medium-5 columns"
    >
      <label>
        Species:
      </label>
      <div
        className=" css-2b097c-container"
        onKeyDown={[Function]}
      >
        <div
          className=" css-15r8ym2-Control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className=" css-1hwfws3"
          >
            <div
              className=" css-1uccc91-singleValue"
            >
              Homo sapiens
            </div>
            <div
              className="css-b8ldur-Input"
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-autocomplete="list"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  disabled={false}
                  id="react-select-4-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  spellCheck="false"
                  style={
                    Object {
                      "background": 0,
                      "border": 0,
                      "boxSizing": "content-box",
                      "color": "inherit",
                      "fontSize": "inherit",
                      "label": "input",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "1px",
                    }
                  }
                  tabIndex="0"
                  type="text"
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              className="sc-gzVnrw coZbQ"
              cx={[Function]}
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="small-12 medium-5 columns"
    >
      <label>
        Inferred cell types:
      </label>
      <div
        className=" css-2b097c-container"
        onKeyDown={[Function]}
      >
        <div
          className=" css-15r8ym2-Control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className=" css-1hwfws3"
          >
            <div
              className=" css-1uccc91-singleValue"
            >
              beta cell
            </div>
            <div
              className="css-b8ldur-Input"
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-autocomplete="list"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  disabled={false}
                  id="react-select-5-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  spellCheck="false"
                  style={
                    Object {
                      "background": 0,
                      "border": 0,
                      "boxSizing": "content-box",
                      "color": "inherit",
                      "fontSize": "inherit",
                      "label": "input",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "1px",
                    }
                  }
                  tabIndex="0"
                  type="text"
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              className="sc-gzVnrw coZbQ"
              cx={[Function]}
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="small-12 medium-2 columns"
      style={
        Object {
          "paddingTop": "25px",
        }
      }
    >
      <button
        className="button"
        onClick={[Function]}
      >
        Search
      </button>
    </div>
  </div>
  <div
    style={
      Object {
        "paddingBottom": "25px",
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="columns large-9 large-centered"
      >
        <div
          className="callout alert small"
        >
          <h5>
            Oops!
          </h5>
          <p>
            There was a problem rendering this component.
            <br />
            If the error persists, in order to help us debug the issue, please copy the URL and this message and send it to us via 
            <a
              href="https://www.ebi.ac.uk/support/gxasc"
            >
              the EBI Support & Feedback system
            </a>
            :
          </p>
          <code>
            TypeError: Cannot read property 'nodeName' of null â€“ [object Object]
          </code>
        </div>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "alignItems": "center",
        "background": "rgba(255,255,255,0.8)",
        "bottom": 0,
        "display": "none",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <div
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      <p>
        Loading, please wait...
      </p>
      <img
        src="test-file-stub"
      />
    </div>
  </div>
</div>
`;
